airflow/providers/google/cloud/hooks/compute_ssh.py

```diff
@@ -0,0 +1,250 @@
+                              http=http_authorized, cache_discovery=False)
+        return self.conn
+
+    def get_sshclient(self,
+                      private_key_file: str,
+                      hostname: str,
+                      username: str,
+                      strict_host_checking: bool = True
+                      ) -> paramiko.SSHClient:
+        """
+        Get SSHClient
+
+        :param private_key_file: The filename of the private key
+        :type private_key_file: str
+        :param hostname: The hostname of the target instance
+        :type hostname: str
+        :param username: The POXIS username from the login profile of the user account
+        :type username: str
+        :param strict_host_checking: Whether to verify key
+        :type strict_host_checking: bool
+        """
+
+        self.log.info("Creating remote connection to host")
+        client = paramiko.SSHClient()
+        if not strict_host_checking:
+            self.log.warning('No Host Key Verification. This wont protect '
+                             'against Man-In-The-Middle attacks')
+            # Default is RejectPolicy
+            client.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+
+        client.connect(hostname=hostname,
+                       username=username,
+                       key_filename=private_key_file)
+        return client
+
+    def get_service_account(self) -> str:
+        """
+        Get the service account id
+        """
+        account: str = self._get_credentials().service_account_email
+        if not account.startswith('users/'):
+            account = 'users/' + account
+        return account
+
+    def get_username(self) -> str:
+        """
+        Using the OS Login API, get the POSIX username from the login profile
+        for the service account.
+
+        """
+        conn = self.get_conn()
+        account = self.get_service_account()
+        profile = conn.users().getLoginProfile(name=account).execute_in_compute()  # pylint: disable=no-member
+        username: str = profile.get('posixAccounts')[0].get('username')
+        return username
+
+    @GoogleBaseHook.fallback_to_default_project_id
+    def get_hostname(self, project_id: Optional[str], instance: str, zone: str,
+                     hostname: Optional[str] = None) -> str:
+        """
+        Create the target instance hostname
+
+        :param project_id: The project ID of the instance
+        :type project_id: Optional[str]
+        :param instance: The name of the target instance to connect to
+        :type instance: str
+        :param zone: The zone of the target instance
+        :param zone: str
+        :param hostname: The hostname of the target instance. If not provided,
+            it is generated from zone, instance name and project id
+        :type hostname: str
+        """
+        if not hostname:
+            hostname = f'{instance}.{zone}.c.{project_id}.internal'
```

Comment:
Hostname is a bit complicated in this case, but we definitely shouldn't use a domain name, but IP addresses.

If user passes hostname, we should use the value provided by user
If user turns on IAP tunnel, we should use a private IP address.
If user enforce private IP (`use_internal_ip` option`, then we should use private IP
Otherwise, we should use the public IP address.

Public IP address is a first non-empty value in `networkInterfaces[].accessConfigs[].natIP` attribute of [Instance object](https://cloud.google.com/compute/docs/reference/rest/v1/instances/get).
Private IP address is a first non-empty value in `networkInterfaces[].networkIP` attribute of [Instance object](https://cloud.google.com/compute/docs/reference/rest/v1/instances/get).